### PR TITLE
Migrate Country list to symfony

### LINKED ISSUE #10604: Migrate "International > Locations > Countries" page
Part of Symfony migration project
Type: page

## Status

Todo:
- [x] Migrate listing 
- [ ] Form todo
Capture d’écran 2023-07-06 à 15 52 23

- [ ] [Add a modal like on product list (Catalog > Products) to confirm or cancel delete in bulk action](https://github.com/PrestaShop/PrestaShop/issues/14462)
- [ ] [Add a modal to confirm or cancel delete action for one row](https://github.com/PrestaShop/PrestaShop/issues/17082)
- [ ] #19372 - Checkboxes for multistore
- [ ] #19324 - Specific settings dropdown for multistore

#### Controller / routing
- [ ] All controller actions are migrated and protected with `@AdminSecurity` annotations when eligible and `@DemoRestricted` annotations when eligible
- [ ] All links targeting this page are updated: either using _legacy_link or updating the code
- [ ] The new SF routes have `_legacy_link` and `_legacy_parameters` provided and explained
- [ ] Controller catches Domain Exception and translate them nicely
#### Template
- [ ] Twig templates must contain relevant blocks to allow partial extension by modules
- [ ] Twig forms must have `form_rest()` and `form_errors()`
- [ ] JS being used on the page can be used as enabled extensions and use a PageMap for selectors
- [ ] Use `@PrestaShop` instead of `PrestaShopBundle` for template root path aliasing
##### Conventions
- [ ] The new controller actions comply with our naming convention
- [ ] The new SF routes names and paths comply with our naming convention
- [ ] The new Twig template names comply actions with our naming convention
- [ ] Created files (controller, JS assets, CSS assets and twig views) must be in the right folder following the convention
- [ ] CSS being added must follow theme conventions

#### Core layer
- [ ] Commands and Queries only use native types but internally use VOs
- [ ] All Handlers must have an Interface
##### Conventions
- [ ] The created SF service names comply with our naming convention
- [ ] The created files must follow the folder convention (Domain, Core, Adapter...)

#### Polishing
- [ ] The legacy controller and templates files are gone
- [ ] If PR includes a new concept/mechanism, documentation must be written
- [ ] JS and CSS assets must be compiled and up-to-date
- [ ] License headers must be up-to-date
- [ ] New hooks have been added to the hooks list and the SQL updates
- [ ] Performance must be correct: page loading time, filtering/sorting/searching response time, actions response time

#### Testing
- [ ] Commands and Queries are tested with Behat tests
- [ ] Eligible Core classes are covered with unit tests*
- [ ] Controller is covered with survival test
- [ ] (soon) Controller is covered with E2E tests using puppeeter
- [ ] If we break E2E tests because IDs change, we must update the E2E tests

*Eligible = easy to isolate and unit tests provide values (for example testing getters/setters is useless)

### PR DESCRIPTION
**Pull request by margud**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Migration of /international/locations/countries list. This PR only contains grid and search. No actions implemented yet.
| Type?             | refacto
| Category?         | BO 
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes [#10604](https://github.com/PrestaShop/PrestaShop/issues/10604)
| How to test?      | You can access controller by using route in url: improve/international/countries. It should look just like legacy but without actions, only filter is implemented.

### CODE REVIEW

File: src/Core/Grid/Query/CountryQueryBuilder.php
Comment (zuk3975) [inline src/Core/Grid/Query/CountryQueryBuilder.php]:
```suggestion
     * @param int[] $contextShopIds
```

File: src/PrestaShopBundle/Controller/Admin/Improve/International/CountryController.php
Comment (zuk3975) [inline src/PrestaShopBundle/Controller/Admin/Improve/International/CountryController.php]: try using CommonController:searchGridAction instead

File: src/Core/Grid/Query/CountryQueryBuilder.php
Comment (zuk3975) [inline src/Core/Grid/Query/CountryQueryBuilder.php]: not sure if we should use c.* or only select actually used fields? ping @jolelievre (if we select all fields it may sometimes help module developers to add some fields without needing to modify query builder)

File: src/PrestaShopBundle/Resources/config/routing/admin/improve/international/country.yml
Comment (Progi1984) [inline src/PrestaShopBundle/Resources/config/routing/admin/improve/international/country.yml]:
```suggestion
```

File: src/PrestaShopBundle/Controller/Admin/Improve/International/CountryController.php
Comment (Progi1984) [inline src/PrestaShopBundle/Controller/Admin/Improve/International/CountryController.php]:
```suggestion
 */

declare(strict_types=1);

namespace PrestaShopBundle\Controller\Admin\Improve\International;
```

File: src/PrestaShopBundle/Controller/Admin/Improve/International/CountryController.php
Comment (Progi1984) [reply src/PrestaShopBundle/Controller/Admin/Improve/International/CountryController.php]: :+1: with @zuk3975

File: src/Core/Search/Filters/CountryFilters.php
Comment (Progi1984) [inline src/Core/Search/Filters/CountryFilters.php]:
```suggestion
class CountryFilters extends Filters
```

File: src/Core/Search/Filters/CountryFilters.php
Comment (Progi1984) [inline src/Core/Search/Filters/CountryFilters.php]:
```suggestion
 */

declare(strict_types=1);

namespace PrestaShop\PrestaShop\Core\Search\Filters;
```

File: src/Core/Grid/Query/CountryQueryBuilder.php
Comment (Progi1984) [inline src/Core/Grid/Query/CountryQueryBuilder.php]:
```suggestion
class CountryQueryBuilder extends AbstractDoctrineQueryBuilder
```

File: src/Core/Grid/Definition/Factory/CountryGridDefinitionFactory.php
Comment (Progi1984) [inline src/Core/Grid/Definition/Factory/CountryGridDefinitionFactory.php]:
```suggestion
class CountryGridDefinitionFactory extends AbstractGridDefinitionFactory
```

**Review by Progi1984 — COMMENTED** [review]
Some feedbacks

File: src/Core/Grid/Definition/Factory/CountryGridDefinitionFactory.php
Comment (Julievrz) [inline src/Core/Grid/Definition/Factory/CountryGridDefinitionFactory.php]:
```suggestion
        return $this->trans('Country', [], 'Admin.Global');
```

File: src/Core/Grid/Definition/Factory/CountryGridDefinitionFactory.php
Comment (Julievrz) [inline src/Core/Grid/Definition/Factory/CountryGridDefinitionFactory.php]:
```suggestion
                    ->setName($this->trans('ISO Code', [], 'Admin.International.Feature'))
```

File: src/Core/Grid/Definition/Factory/CountryGridDefinitionFactory.php
Comment (Julievrz) [inline src/Core/Grid/Definition/Factory/CountryGridDefinitionFactory.php]:
```suggestion
                    ->setName($this->trans('Call prefix', [], 'Admin.International.Feature'))
```

File: src/Core/Grid/Query/CountryQueryBuilder.php
Comment (zuk3975) [reply src/Core/Grid/Query/CountryQueryBuilder.php]: just to clarify for @margud, we used to finalize all classes, thats why you can still see lots of these examples, but recently it was decided to avoid finalizing classes to allow extending more from modules. So don't use `final` anymore :smile: (probably there are some exceptions where it could still be used, but don't bother, high chances you wont ever need to use them in ps)

File: src/PrestaShopBundle/Resources/config/routing/admin/improve/international/country.yml
Comment (margud) [reply src/PrestaShopBundle/Resources/config/routing/admin/improve/international/country.yml]: Added this todo to make sure we dont forget to remove it later on when page is finished. Now we dont want to override the controller because its not finished.

File: src/Core/Search/Filters/CountryFilters.php
Comment (margud) [reply src/Core/Search/Filters/CountryFilters.php:35]: why should we remove final here? All filters are final class.

File: admin-dev/themes/new-theme/js/pages/country/index.ts
Comment (Progi1984) [inline admin-dev/themes/new-theme/js/pages/country/index.ts]: Please update your license headers

File: src/Core/Grid/Query/CountryQueryBuilder.php
Comment (Progi1984) [inline src/Core/Grid/Query/CountryQueryBuilder.php]:
```suggestion
    protected $searchCriteriaApplicator;

    /**
     * @var int[]
     */
    protected $contextShopIds;
```

File: src/Core/Grid/Query/CountryQueryBuilder.php
Comment (Progi1984) [inline src/Core/Grid/Query/CountryQueryBuilder.php]:
```suggestion
    protected function getQueryBuilder(SearchCriteriaInterface $searchCriteria): QueryBuilder
```

File: src/Core/Grid/Query/CountryQueryBuilder.php
Comment (Progi1984) [inline src/Core/Grid/Query/CountryQueryBuilder.php]:
```suggestion
    protected function applyFilters(QueryBuilder $builder, SearchCriteriaInterface $criteria): void
```

File: tests/Integration/PrestaShopBundle/Controller/Admin/Improve/International/CountryControllerTest.php
Comment (Progi1984) [inline tests/Integration/PrestaShopBundle/Controller/Admin/Improve/International/CountryControllerTest.php]:
```suggestion
    public function testCreate(int $initialEntityCount): void
```

File: tests/Integration/PrestaShopBundle/Controller/Admin/Improve/International/CountryControllerTest.php
Comment (Progi1984) [inline tests/Integration/PrestaShopBundle/Controller/Admin/Improve/International/CountryControllerTest.php]:
```suggestion
    public function testEdit(int $countryId): void
```

File: tests/Integration/PrestaShopBundle/Controller/Admin/Improve/International/CountryControllerTest.php
Comment (Progi1984) [inline tests/Integration/PrestaShopBundle/Controller/Admin/Improve/International/CountryControllerTest.php]:
```suggestion
    public function testFilters(int $countryId): void
```

**Review by Progi1984 — COMMENTED** [review]
Some last feedbacks

**Comment by margud** [comment]
> Some last feedbacks

Fixed

File: src/Core/Grid/Query/CountryQueryBuilder.php
Comment (zuk3975) [inline src/Core/Grid/Query/CountryQueryBuilder.php]:
```suggestion
    /**
     * @var DoctrineSearchCriteriaApplicatorInterface
     */
```

i think we use it this way everywhere :thinking:

File: src/Core/Grid/Query/CountryQueryBuilder.php
Comment (zuk3975) [inline src/Core/Grid/Query/CountryQueryBuilder.php]: same as above

File: src/Core/Grid/Definition/Factory/CountryGridDefinitionFactory.php
Comment (jolelievre) [inline src/Core/Grid/Definition/Factory/CountryGridDefinitionFactory.php:121]: If it is the only feature missing for the grid in this PR then I think it's worth handling it here. It shouldn't make the scope explode It's one command, one controller action, one JS extension to add, we have enough tools and example to do this one quickly. This way we only have one round of QA to do and the list page is fully functional.

File: src/Core/Grid/Definition/Factory/CountryGridDefinitionFactory.php
Comment (jolelievre) [inline src/Core/Grid/Definition/Factory/CountryGridDefinitionFactory.php:220]: You need to add the row actions here, even if the form part is not implemented what we usually do is create the routes and controllers. Only the controllers are very basic they simply redirect to the legacy page of the form.

File: src/Core/Grid/Query/CountryQueryBuilder.php
Comment (jolelievre) [inline src/Core/Grid/Query/CountryQueryBuilder.php]: Are you sure you need a groupBy here? This should work without it I think

And if it is really needed for some reason it should be part of the `getQueryBuilder` or the the count could be different Although in this case it would work thanks to the DISTINCT in the count but it's still safer to have the same base for the query in both count and result methods

File: src/Core/Grid/Query/CountryQueryBuilder.php
Comment (jolelievre) [inline src/Core/Grid/Query/CountryQueryBuilder.php:99]: I'm terrible at join subtilities so excuse my question if it's dumb 😅
But why an innerJoin here instead of a leftJoin?

File: src/PrestaShopBundle/Resources/config/routing/admin/improve/international/country.yml
Comment (jolelievre) [reply src/PrestaShopBundle/Resources/config/routing/admin/improve/international/country.yml]: We usually know we have to uncomment those because it's what actually mark the page as released But an extra comment for clarity is not blocking here

File: tests/Integration/PrestaShopBundle/Controller/Admin/Improve/International/CountryControllerTest.php
Comment (jolelievre) [inline tests/Integration/PrestaShopBundle/Controller/Admin/Improve/International/CountryControllerTest.php]: Why not, it seems like filters are implemented so you should be able to test them based on the fixtures data

**Review by jolelievre — COMMENTED** [review]
Thanks @margud 

A few feedback, also should we handle the whole list feature in this PR or do you plan to finish the missing part in another one?

File: src/Core/Grid/Definition/Factory/CountryGridDefinitionFactory.php
Comment (jolelievre) [reply src/Core/Grid/Definition/Factory/CountryGridDefinitionFactory.php:121]: Intermediate PR so ok to do it later

File: src/Core/Grid/Definition/Factory/CountryGridDefinitionFactory.php
Comment (jolelievre) [reply src/Core/Grid/Definition/Factory/CountryGridDefinitionFactory.php:220]: Please add a todo here, the row actions must be added at some point even if the form is not done we can create controllers that simply redirect to legacy page

File: src/Core/Grid/Definition/Factory/CountryGridDefinitionFactory.php
Comment (jolelievre) [inline src/Core/Grid/Definition/Factory/CountryGridDefinitionFactory.php:266]: Add a todo here as well, we have several bulk actions to handle here

**Review by jolelievre — COMMENTED** [review]
I didi a second review with the intermediate PR in mind.
So we need to add a few todos to make sure nothing is forgotten, and I'm still wary about my comments in the builder.

Finally correct me if I'm wrong but I believe the filters are testable in the function tests? Since we have a bunch of countries in the fixtures data we don't need the form to create some.

**Review by Progi1984 — APPROVED** [review]

File: src/Core/Grid/Query/CountryQueryBuilder.php
Comment (zuk3975) [reply src/Core/Grid/Query/CountryQueryBuilder.php]: Screenshot from 2022-05-23 17-53-27

Sorry, Progi, but I think it applies the same as in some of my PR's.
As @matks said its not (yet) the standard and it shouldn't be asked to be changed everywhere.

Comment (Progi1984) [reply src/Core/Grid/Query/CountryQueryBuilder.php]: @zuk3975 Sorry. But I disagree with matks. Contributions to the project must allow extensibility and inheritance so my comments are always valid.

**Review by Progi1984 — DISMISSED** [review]

File: src/Core/Grid/Query/CountryQueryBuilder.php
Comment (margud) [reply src/Core/Grid/Query/CountryQueryBuilder.php:99]: inner join because we expect it to always exists, left join is used when it might be missing. Also i m using zone controller as example and i see that it also used same logic.

File: src/PrestaShopBundle/Controller/Admin/Improve/International/CountryController.php
Comment (jolelievre) [inline src/PrestaShopBundle/Controller/Admin/Improve/International/CountryController.php]: You need to use the `$this->redirect` Symfony method here instead of the legacy one, or as you can see PHPStan will be quite unhappy ^^

File: src/PrestaShopBundle/Controller/Admin/Improve/International/CountryController.php
Comment (jolelievre) [inline src/PrestaShopBundle/Controller/Admin/Improve/International/CountryController.php:75]: Can you add the admin security here to avoid forgetting it later `@AdminSecurity("is_granted('update', request.get('_legacy_controller'))")`

File: src/PrestaShopBundle/Controller/Admin/Improve/International/CountryController.php
Comment (jolelievre) [inline src/PrestaShopBundle/Controller/Admin/Improve/International/CountryController.php]:
```suggestion
            Context::getContext()->link->getAdminLink(
```

Comment (jolelievre) [reply src/PrestaShopBundle/Controller/Admin/Improve/International/CountryController.php]: We avoid using FQCN even for legacy objects which have no namespace (especially for them actually since they are the ones that are added by the IDEs automatically the most 😅 )
Instead, you should add a `use Context;` on top of the file

File: src/PrestaShopBundle/Resources/config/routing/admin/improve/international/country.yml
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/config/routing/admin/improve/international/country.yml]:
```suggestion
# todo uncomment all legacy links when page is finished.
```

Comment (jolelievre) [reply src/PrestaShopBundle/Resources/config/routing/admin/improve/international/country.yml]: This will avoid having this warning message on all the project PRs

File: src/PrestaShopBundle/Resources/config/routing/admin/improve/international/country.yml
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/config/routing/admin/improve/international/country.yml]:
```suggestion
# admin_zones_delete:
```

File: src/PrestaShopBundle/Resources/config/routing/admin/improve/international/country.yml
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/config/routing/admin/improve/international/country.yml]:
```suggestion
#   _legacy_link: AdminCountry:updateCountry
```

Comment (jolelievre) [reply src/PrestaShopBundle/Resources/config/routing/admin/improve/international/country.yml]: SAme here try not to introduce some warning in the YAML files, because once the PR is merged every pending PRs will also have them which can be annoying

Comment (jolelievre) [reply src/PrestaShopBundle/Resources/config/routing/admin/improve/international/country.yml]: You can run the linter locally with `php bin/console lint:yaml src` to make sure everything is ok before pushing it It's faster than waiting for the CI from github

**Review by jolelievre — COMMENTED** [review]
Thanks @margud 

All good for me, the remaining comments are to make the CI happy mainly ^^

File: src/PrestaShopBundle/Resources/config/routing/admin/improve/international/country.yml
Comment (margud) [reply src/PrestaShopBundle/Resources/config/routing/admin/improve/international/country.yml]: I don't really get the problem here. I can't see any warnings, its how every other yml is commenting and if i remove the space after uncommenting the line it will have bad spacing so not sure why we need to remove the space here or where i can see the warnings. The command lint:yaml doesn't say anything about this line.

Comment (jolelievre) [reply src/PrestaShopBundle/Resources/config/routing/admin/improve/international/country.yml]: Actually I looked at the wrong command it's actually `yamllint -c .github/workflows/yamllint/.yamllint.yml src` which had this warning but it seems all good now

**Review by jolelievre — APPROVED** [review]
Thanks @margud

**Review by atomiix — APPROVED** [review]

**Comment by jolelievre** [comment]
This is a partial PR for the country list, some additional modifications are still needed and the page is hidden so no need for QA right now it will be tested when the whole page is finished.

**Comment by jolelievre** [comment]
Thank you @margud

**Comment by Robin-Fischer-PS** [comment]
Hi @jolelievre ,

We usually do some QA even if the page is not finished (Product V2 was an exception) :)

**Pull request merged**
